src/Features/Core/Portable/Completion/CompletionServiceWithProviders.ProviderSource.cs

```diff
@@ -0,0 +1,148 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Completion.Providers;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion
+{
+    public abstract partial class CompletionServiceWithProviders
+    {
+        private class ProviderSource : IEqualityComparer<ImmutableHashSet<string>>
+        {
+            private readonly CompletionServiceWithProviders _service;
+            private IEnumerable<Lazy<CompletionProvider, CompletionProviderMetadata>>? _importedProviders;
+
+            private readonly object _gate = new();
+            private readonly Dictionary<string, CompletionProvider?> _nameToProvider = new();
+            private readonly Dictionary<ImmutableHashSet<string>, Task<ImmutableArray<CompletionProvider>>> _rolesToProviders;
+
+            public ProviderSource(CompletionServiceWithProviders service)
+            {
+                _service = service;
+                _rolesToProviders = new(this);
+            }
+
+            public ImmutableArray<CompletionProvider> GetProviders(ImmutableHashSet<string> roles, bool waitUntilAvaialble = false)
+            {
```

Comment:
❔ How is this first invoked? Can we make sure that one call to this occurs immediately after a document is opened and/or `roles` changes? It's really frustrating how completion for unimported types/members never works the first time it's used in a document, and the worst thing possible here would be extending that behavior to all completion operations.

Reply:
I believe this is called first when we try to decide whether to trigger a completion (in `ShouldTriggerCompletion`).  I guess can add a timeout instead of return immediately to mitigate this. Otherwise, adding something that listen to workspace event and warm up cache for completion seems reasonable.

Reply:
A short timeout is added before we bail